### DISCUSSION TITLE
Макросы для работы с выделением и для навигации по папкам (Macros for working with selection and for navigating through folders)

### DISCUSSION THREAD

**Discussion by michel-nk** [elfmz/far2l#2266] [Show and tell]
За многие годы использования ФАРа (ещё под виндой, конечно) привык к набору макросов, которые сильно ускоряют выполнение некоторых операций, часто встречающихся в моих рабочих сценариях. Адаптировал их под Linux и решил поделиться ими с сообществом.

1. Открыть **текущую папку** на пассивной панели (синхронизировать панели):
```
[KeyMacros/Shell/AltShiftPgUp]
DisableOutput=0x1
Sequence=$If (APanel.FilePanel) panel.setpath(1, APanel.Path) $End
```

2. Открыть **папку под курсором** на пассивной панели:
```
[KeyMacros/Shell/AltShiftPgDn]
DisableOutput=0x1
NoFiles=0x1
Sequence=$If (APanel.Folder) panel.setpath(1, APanel.Path + "/" + APanel.Current) $End
```

3. Пара макросов, позволяющая перемещаться по списку папок вверх или вниз с просмотром их содержимого на пассивной панели. Фактически это обёртка вокруг второго макроса, добавляющая ещё и навигацию по панели:
```
[KeyMacros/Shell/AltShiftDown]
DisableOutput=0x1
NoFiles=0x1
Sequence=down $If (APanel.Folder) panel.setpath(1, APanel.Path + "/" + APanel.Current) $End

[KeyMacros/Shell/AltShiftUp]
DisableOutput=0x1
NoFiles=0x1
Sequence=up $If (APanel.Folder) panel.setpath(1, APanel.Path + "/" + APanel.Current) $End
```

4. Пара макросов, позволяющая **копировать или переносить отмеченные объекты в папку под курсором**. Например, отметили несколько объектов на панели, стали курсором на какую-то папку в этой же панели (в том числе на папку ..) и вызвали макрос. Отмеченные объекты копируются/перемещаются в эту папку. Удобно для раскладывания файлов по подпапкам в пределах одной папки без задействования соседней панели (она ведь тоже всегда для чего-то нужна и покидать её не хочется).
```
[KeyMacros/Shell/AltShiftF6]
DisableOutput=0x1
NoFiles=0x1
NoSendKeysToPlugins=0x1
Sequence=F6 Del . / ShiftEnter / Enter

[KeyMacros/Shell/AltShiftF5]
DisableOutput=0x1
NoFiles=0x1
NoSendKeysToPlugins=0x1
Sequence=F5 Del . / ShiftEnter / Enter
```

**Теперь макросы для работы с отмеченными объектами**

5. Пара макросов, позволяющая перемещать курсор к следующему/предыдущему отмеченному объекту на панели:
```
[KeyMacros/Shell/AltDown]
DisableOutput=0x1
Sequence=$If ((APanel.FilePanel) && (APanel.Selected)) %i = APanel.curPos + 1; $While (%i <= APanel.itemCount) $If (panel.item(0, %i,8)) panel.setposidx(0, %i); $Exit $End %i=%i+1; $End $End

[KeyMacros/Shell/AltUp]
DisableOutput=0x1
Sequence=$If ((APanel.FilePanel) && (APanel.Selected)) %i = APanel.curPos - 1; $While (%i >= 1) $If (panel.item(0, %i,8)) panel.setposidx(0, %i); $Exit $End %i=%i-1; $End $End
```

6. Макрос, отмечающий на пассивной панели объекты, отмеченные на активной. Например, отметили несколько файлов/папок на активной панели и хотим быстро поискать аналогичные на пассивной панели. Этот макрос именно это и сделает. При наличии каких-то отмеченных файлов на пассивной панели макрос задаст вопрос и потом может или сначала сбросить существующую отметку или объединить с ней:
```
[KeyMacros/Shell/AltShiftS]
DisableOutput=0x1
Sequence=$If ((APanel.FilePanel) && (APanel.Selected) && (PPanel.FilePanel)) %savedPos = PPanel.curPos; $If (PPanel.Selected) $If (msgbox("На целевой панели уже есть выделение", "Сбросить текущее выделение на целевой панели (иначе выделения будут объединены)?", 0x00040000) == 1) panel.select(1, 0); $End $End %i = 1; $While (%i <= APanel.itemCount) $If (panel.item(0, %i,8)) %nextFName = panel.item(0, %i, 0); %newPos = panel.setpos(1, %nextFName); $If (%newPos > 0) panel.select(1, 1, 1, 0); $End $End %i=%i+1; $End %savedPos = panel.setposidx(1, %savedPos); $End
```

7. И, наконец, самый крупный макрос - отметить объекты по списку из буфера обмена. Простейший (и основной для меня) способ формирования такого списка - копирование текущей отметки прямо из ФАРа. Но одновременно поддерживаются списки с разделителем (встроенная возможность макроязыка ФАР). Аналогично предыдущему макросу, опционально может выполнить объединение с существующими отмеченными объектами:
```
[KeyMacros/Shell/AltShiftV]
DisableOutput=0x1
Sequence=$If (APanel.FilePanel) %savedPos = APanel.curPos; %theEndOfStr = chr(0x0A); %clipContent = replace(clip(0), chr(0x0D), ""); $If (len(%clipContent) < 1) $Exit $End $If (APanel.Selected) $If (msgbox("На панели уже есть выделение", "Сбросить текущее выделение (иначе выделения будут объединены)?", 0x00040000) == 1) panel.select(0, 0); $End $End $If ((index(%clipContent, %theEndOfStr) > 0) && (index(%clipContent, ";") < 0) && (index(%clipContent, ",") < 0)) %clipContent = replace(%clipContent, %theEndOfStr, ";"); panel.select(0, 1, 2, %clipContent); $Exit $End $If (index(%clipContent, %theEndOfStr) < 0) panel.select(0, 1, 2, %clipContent); $Exit $End %endStrPos = index(%clipContent, %theEndOfStr); $While (%endStrPos >= 0) %curFileName = subStr(%clipContent, 0 , %endStrPos); %clipLenBefore = len(%clipContent); %clipContent = subStr(%clipContent, %endStrPos + 1, len(%clipContent)); $If (%clipLenBefore <= len(%clipContent)) $Exit $End %filePos = panel.setpos(0, %curFileName); $If (%filePos > 0) panel.select(0, 1, 1, 0); $End %endStrPos = index(%clipContent, %theEndOfStr); $End $If (len(%clipContent) > 0) %filePos = panel.setpos(0, %clipContent); $If (%filePos > 0) panel.select(0, 1, 1, 0); $End $End %savedPos = panel.setposidx(0, %savedPos); $End
```

**И ещё бонусом добавлю одну возможность, которую сделал в пользовательском меню - копирование файлов через системный буфер обмена.** Теперь можно скопировать набор объектов в ФАРе, а вставить его в системном файловом менеджере и наоборот (проверено в Ubuntu (GNOME) и Kubuntu (Plasma)). Разумеется, в последнем случае надо помнить о том, что пути к объектам должны поддерживаться командой rsync т.к. фактическую работу по "вставке из буфера" выполняет именно она.
Для такой интеграции нужно:
1. Доустановить в систему пакет **xclip**
2. В папке ФАРа в профиле пользователя создать скрипт (+ права на исполнение) **copy-to-clipboard.sh** с содержимым:
```
#!/bin/bash
# Первый и единственный параметр - имя файла-списка, передаваемое из ФАРа
# sed - Вставим в начале каждой строки фрагмент file://
sed -E -i 's/^(.)/file\:\/\/\1/' $1
xclip -selection clipboard -t text/uri-list $1
```

3. В папке ФАРа в профиле пользователя создать скрипт (+ права на исполнение) **paste-from-clipboard.sh** с содержимым:
```
#!/bin/bash
# Первый и единственный НЕОБЯЗАТЕЛЬНЫЙ параметр - возможный ключ -u для обновления существующих файлов 
addrPrefix="file://" 

xclip -selection clipboard -o | while read theLine ; do
  theIndex=`expr index "$theLine" "$addrPrefix"`
  if [[ $theIndex = 1 ]]; then
    newLine=${theLine/"$addrPrefix"}
    rsync -a $1 "$newLine" "${PWD}"
  else
    echo "Неожиданный формат адреса в строке:" $theLine;
  fi
done
```
4. Добавить в файл пользовательского меню, например, такие команды:
```
C:  Буфер обмена (системный)
{
C:  Скопировать помеченные объекты
    ~/.config/far2l/copy-to-clipboard.sh !@F!
V:  Вставить объекты (обновление)
    ~/.config/far2l/paste-from-clipboard.sh -u
O:  Вставить объеты (всегда заменять)
    ~/.config/far2l/paste-from-clipboard.sh
--:  
S:  Показать содержимое
    xclip -selection clipboard -o
}
```
Если при этом лень нажимать F2;C;C|V для копирования/вставки, можно навесить какую-нибудь клавиатурную комбинацию.

**Comment by akruphi** [comment]
Круто! Полезные макросы. Утянул ссылкой в русскоязычный вики: https://github.com/akruphi/far2l/wiki/#key_macros
